### PR TITLE
CLIP-1569:  Add configurable access restriction to the deployed applications.

### PR DESCRIPTION
**Pull request by nghazali**
## Pull request description

We have no restriction for accessing the applications after deployment. As part of improving the security and accessibility management, we should add some configurable restriction to limit the public to run the applications. We have similar feature implemented in AWS QuickStart. 
As part of this PR, a new feature will be introduce to allow admins limit the whitelist for running the applications. The default value for the whitelist  is a public access (0.0.0.0/0). 

## Checklist
- [x] I have successful end to end tests run (with & without domain)
- [x] I have added unit tests (if applicable)
- [x] I have user documentation (if applicable)

### CODE REVIEW

**Review by yzha645 — APPROVED** [review]

File: config.tfvars [hunk 1]
```diff
@@ -18,6 +18,11 @@ region = "<REGION>"
 # e.g.: products = ["jira", "confluence"]
 products = ["<LIST_OF_PRODUCTS>"]
 
+# Range of IPs that are allowed to run the applications.
```
Comment (errcode1202) [inline config.tfvars:21]: Maybe

```suggestion
# Range of IPs that are allowed to access the running applications over the World Wide Web.
```

File: config.tfvars [hunk 2]
```diff
@@ -18,6 +18,11 @@ region = "<REGION>"
...
+# By default the deployed applications are publicly accessible (0.0.0.0/0). You can restrict this access by change the
```
Comment (errcode1202) [inline config.tfvars:22]:
```suggestion
# By default the deployed applications are publicly accessible (0.0.0.0/0). You can restrict this access by changing the
```

File: dc-infrastructure.tf [hunk 3]
```diff
@@ -11,6 +11,7 @@ module "base-infrastructure" {
   domain               = var.domain
   namespace            = local.namespace
   eks_additional_roles = var.eks_additional_roles
+  whitelist_cidr       = var.whitelist_cidr
```
Comment (errcode1202) [inline dc-infrastructure.tf:14]: https://hello.atlassian.net/wiki/spaces/CONTENT/pages/542809951/Terms+Whitelist+Blacklist+in+software+Allowlist+Blocklist

File: modules/AWS/ingress/variables.tf [hunk 4]
```diff
@@ -9,3 +9,13 @@ variable "enable_ssh_tcp" {
   type        = bool
   default     = false
 }
+
+variable "loadBalancerSourceRanges" {
+  description = "List of allowed CIDRs to access to the load balancer."
```
Comment (errcode1202) [inline modules/AWS/ingress/variables.tf:14]:
```suggestion
  description = "List of allowed CIDRs that can access the load balancer."
```

File: modules/common/variables.tf [hunk 5]
```diff
@@ -56,4 +56,9 @@ variable "enable_ssh_tcp" {
 variable "eks_additional_roles" {
   description = "Additional roles that have access to the cluster."
   type        = list(object({ rolearn = string, username = string, groups = list(string) }))
-}
\ No newline at end of file
+}
+
+variable "whitelist_cidr" {
+  description = "List of CIDRs allow to access to application(s)."
```
Comment (errcode1202) [inline modules/common/variables.tf:62]:
```suggestion
  description = "List of CIDRs allowed that have access to the application(s)."
```

File: test/unittest/test_variables.go [hunk 6]
```diff
@@ -245,6 +245,22 @@ var DbVariableWithInvalidDBMasterPassword = map[string]interface{}{
 	"db_master_password":  "123@",
 }
 
+// Ingress
+
```
Comment (errcode1202) [inline test/unittest/test_variables.go:249]: Remove carriage return

File: test/unittest/test_variables.go [hunk 7]
```diff
@@ -245,6 +245,22 @@ var DbVariableWithInvalidDBMasterPassword = map[string]interface{}{
...
+}
+
+var IngressValidVariablesValue = map[string]interface{}{
+	"loadBalancerSourceRanges": []string{
+		"10.12.0.0/16",
+		"10.13.1.1/32",
+	},
+}
+
 // Bitbucket
```
Comment (errcode1202) [inline test/unittest/test_variables.go:265]: Remove carriage return

**Review by errcode1202 — CHANGES_REQUESTED** [review]

File: modules/common/main.tf [hunk 8]
```diff
@@ -28,8 +28,9 @@ module "ingress" {
   depends_on = [module.eks]
 
   # inputs
-  ingress_domain = local.ingress_domain
-  enable_ssh_tcp = var.enable_ssh_tcp
+  ingress_domain           = local.ingress_domain
+  enable_ssh_tcp           = var.enable_ssh_tcp
+  loadBalancerSourceRanges = var.whitelist_cidr
```
Comment (nanux) [inline modules/common/main.tf:33]:
```suggestion
  load_balancer_access_ranges = var.whitelist_cidr
```

File: test/unittest/ingress_test.go [hunk 9]
```diff
@@ -46,7 +47,9 @@ func TestIngressIsCreatedWithDomain(t *testing.T) {
 func TestIngressIsCreatedWithoutDomain(t *testing.T) {
 	t.Parallel()
 
-	tfOptions := GenerateTFOptions(map[string]interface{}{}, t, ingressModule)
+	tfOptions := GenerateTFOptions(map[string]interface{}{
+		"loadBalancerSourceRanges": []string{"0.0.0.0/0"},
```
Comment (nanux) [inline test/unittest/ingress_test.go:51]:
```suggestion
		"load_balancer_access_ranges": []string{"0.0.0.0/0"},
```

File: test/unittest/test_variables.go [hunk 10]
```diff
@@ -245,6 +245,22 @@ var DbVariableWithInvalidDBMasterPassword = map[string]interface{}{
 	"db_master_password":  "123@",
 }
 
+// Ingress
+
+var IngressInvalidVariablesValue = map[string]interface{}{
+	"loadBalancerSourceRanges": []string{
```
Comment (nanux) [inline test/unittest/test_variables.go:251]:
```suggestion
	"load_balancer_access_ranges": []string{
```

File: test/unittest/test_variables.go [hunk 11]
```diff
@@ -245,6 +245,22 @@ var DbVariableWithInvalidDBMasterPassword = map[string]interface{}{
...
+		"10.12.0/16",
+		"10.13.1.1/32",
+	},
+}
+
+var IngressValidVariablesValue = map[string]interface{}{
+	"loadBalancerSourceRanges": []string{
```
Comment (nanux) [inline test/unittest/test_variables.go:258]:
```suggestion
	"load_balancer_access_ranges": []string{
```

File: modules/common/variables.tf (hunk 5 above)
Comment (nanux) [reply modules/common/variables.tf:62]:
```
"List of CIDRs allowed accessing the application(s)."
```
?

File: variables.tf [hunk 12]
```diff
@@ -107,6 +107,17 @@ variable "eks_additional_roles" {
   type        = list(object({ rolearn = string, username = string, groups = list(string) }))
 }
 
+variable "whitelist_cidr" {
+  description = "List of CIDRs that are allowed access to the application(s)."
```
Comment (nanux) [inline variables.tf:111]:
```suggestion
  description = "List of CIDRs allowed accessing the application(s)."
```

File: test/unittest/ingress_test.go [hunk 13]
```diff
@@ -68,3 +71,24 @@ func TestIngressIsCreatedWithoutDomain(t *testing.T) {
...
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Invalid CIDR.")
+}
+
+func TestIngressVariablesPopulatedWithValidValue(t *testing.T) {
+	t.Parallel()
+	tfOptions := GenerateTFOptions(IngressValidVariablesValue, t, ingressModule)
+
+	plan := terraform.InitAndPlanAndShowWithStruct(t, tfOptions)
+	loadBalancerSourceRanges := plan.RawPlan.Variables["loadBalancerSourceRanges"].Value
```
Comment (nanux) [inline test/unittest/ingress_test.go:89]:
```suggestion
	loadBalancerSourceRanges := plan.RawPlan.Variables["load_balancer_access_ranges"].Value
```

File: test/unittest/ingress_test.go [hunk 14]
```diff
@@ -13,8 +13,9 @@ func TestIngressIsCreatedWithDomain(t *testing.T) {
 	t.Parallel()
 
 	tfOptions := GenerateTFOptions(map[string]interface{}{
-		"ingress_domain": "test.deplops.com",
-		"enable_ssh_tcp": true,
+		"ingress_domain":           "test.deplops.com",
+		"enable_ssh_tcp":           true,
+		"loadBalancerSourceRanges": []string{"0.0.0.0/0"},
```
Comment (nanux) [inline test/unittest/ingress_test.go:18]:
```suggestion
		"load_balancer_access_ranges": []string{"0.0.0.0/0"},
```

File: modules/AWS/ingress/variables.tf [hunk 15]
```diff
@@ -9,3 +9,13 @@ variable "enable_ssh_tcp" {
...
+  type        = list(string)
+  validation {
+    condition = alltrue([
+    for cidr in var.loadBalancerSourceRanges : can(regex("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])/([0-9]|1[0-9]|2[0-9]|3[0-2])$", cidr))])
```
Comment (nanux) [inline modules/AWS/ingress/variables.tf:18]:
```suggestion
    for cidr in var.load_balancer_access_ranges : can(regex("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])/([0-9]|1[0-9]|2[0-9]|3[0-2])$", cidr))])
```

File: modules/AWS/ingress/variables.tf [hunk 16]
```diff
@@ -9,3 +9,13 @@ variable "enable_ssh_tcp" {
   type        = bool
   default     = false
 }
+
+variable "loadBalancerSourceRanges" {
```
Comment (nanux) [inline modules/AWS/ingress/variables.tf:13]:
```suggestion
variable "load_balancer_access_ranges" {
```

**Review by nanux — CHANGES_REQUESTED** [review]

File: modules/AWS/ingress/main.tf [hunk 17]
```diff
@@ -65,6 +65,8 @@ resource "helm_release" "ingress" {
           # https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-typeloadbalancer
           externalTrafficPolicy = "Local"
 
+          loadBalancerSourceRanges = var.loadBalancerSourceRanges
```
Comment (nanux) [inline modules/AWS/ingress/main.tf:68]:
```suggestion
          loadBalancerSourceRanges = var.load_balancer_access_ranges
```

File: test/unittest/test_variables.go (hunk 6 above)
Comment (nghazali) [reply test/unittest/test_variables.go:249]: This is following the same pattern as other sections.

File: test/unittest/test_variables.go (hunk 7 above)
Comment (nghazali) [reply test/unittest/test_variables.go:265]: The same as above comment.

**Review by errcode1202 — APPROVED** [review]

**Review by nanux — APPROVED** [review]

**Pull request merged**